x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/JobManager.java

```diff
@@ -212,45 +213,43 @@ private void getJobFromClusterState(String jobId, ActionListener<Job> jobListene
      * @param jobsListener The jobs listener
      */
     public void expandJobs(String expression, boolean allowNoJobs, ActionListener<QueryPage<Job>> jobsListener) {
-        Map<String, Job> clusterStateJobs = expandJobsFromClusterState(expression, clusterService.state());
         ExpandedIdsMatcher requiredMatches = new ExpandedIdsMatcher(expression, allowNoJobs);
-        requiredMatches.filterMatchedIds(clusterStateJobs.keySet());
-
-        // If expression contains a group Id it has been expanded to its
-        // constituent job Ids but Ids matcher needs to know the group
-        // has been matched
-        Set<String> groupIds = clusterStateJobs.values().stream()
-                .filter(job -> job.getGroups() != null).flatMap(j -> j.getGroups().stream()).collect(Collectors.toSet());
-        requiredMatches.filterMatchedIds(groupIds);
 
         jobConfigProvider.expandJobsWithoutMissingcheck(expression, false, ActionListener.wrap(
                 jobBuilders -> {
-                    // Check for duplicate jobs
-                    for (Job.Builder jb : jobBuilders) {
-                        if (clusterStateJobs.containsKey(jb.getId())) {
-                            jobsListener.onFailure(new IllegalStateException("Job [" + jb.getId() + "] configuration " +
-                                    "exists in both clusterstate and index"));
-                            return;
-                        }
-                    }
-
                     Set<String> jobAndGroupIds = new HashSet<>();
 
                     // Merge cluster state and index jobs
                     List<Job> jobs = new ArrayList<>();
                     for (Job.Builder jb : jobBuilders) {
                         Job job = jb.build();
                         jobAndGroupIds.add(job.getId());
+                        // If expression contains a group Id it has been expanded to its
+                        // constituent job Ids but Ids matcher needs to know the group
+                        // has been matched
                         jobAndGroupIds.addAll(job.getGroups());
                         jobs.add(job);
                     }
 
+                    // Duplicate configs existing in both the clusterstate and index documents are ok
+                    // this may occur during migration of configs.
+                    // Prefer the index configs and filter duplicates.
+                    Map<String, Job> clusterStateJobs = expandJobsFromClusterState(expression, clusterService.state());
+                    for (String clusterStateJobId : clusterStateJobs.keySet()) {
+                        boolean isDuplicate = jobs.stream().anyMatch(job -> job.getId().equals(clusterStateJobId));
```

Comment:
Since `jobAndGroupIds` have groupIds and jobIds in a `Set` already, why can't we use it to check if we have that job in the `index` ?